### PR TITLE
Update our deprecation policy

### PR DESCRIPTION
**Pull request by jverkoey**
Clarified and defined the individual steps of the deprecation policy in more detail, most notably with relation to the expectations around annotating APIs and writing migration guides.

### CODE REVIEW

**Comment by jverkoey** [comment]
Hi all. This is a change to our culture docs, so I'd like to ensure that everyone LGTMs this before it gets merged. I am working on the companion doc for measuring API usage right now, so the go link for that won't work yet.

**Review by codeman7 — APPROVED** [review]

File: contributing/deprecation_policy.md
Comment (romoore) [inline contributing/deprecation_policy.md]: Should we add a step to delete the migration guide? The old API (and docs) will be captured in the relevant releases so I think future releases can drop the guide since the code is no longer available to be migrated from.

File: contributing/deprecation_policy.md
Comment (romoore) [inline contributing/deprecation_policy.md]: Nit: formatting/line length for this file.

File: contributing/deprecation_policy.md
Comment (romoore) [inline contributing/deprecation_policy.md]: Nit/grammar: s/Generally-speaking/generally speaking/

File: contributing/deprecation_policy.md
Comment (romoore) [inline contributing/deprecation_policy.md]: Broken link?

**Review by romoore — DISMISSED** [review]
Could you please fix the link to the measurement doc? It will be helpful for me to understand it as part of the overall deprecation process.

**Comment by jverkoey** [comment]
[cl/208041473](http://cl/208041473) includes the contents of the measurement doc.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: [cl/208041473](http://cl/208041473) includes the contents of the measurement doc.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Done.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Done.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Done.

**Comment by jverkoey** [comment]
Note to all: I will wait for everyone’s lgtm on this PR before I merge it in. This means your feedback is required - please take a look even if you see that someone else has lgtm’d the PR.

File: contributing/deprecation_policy.md
Comment (romoore) [inline contributing/deprecation_policy.md]: Suggestion: "Measure the internal usage of the API" since we have no mechanism to measure third party use.

**Review by romoore — APPROVED** [review]

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Done.

**Comment by jverkoey** [comment]
Gentle reminder gang: I need an LGTM from everyone before this can be merged.

@randallli 
 @ianegordon
 @jmdetloff 
 @galiak11 
 @yarneo 
 @andrewoverton 
@wenyuzhang666

**Review by yarneo — APPROVED** [review]

File: contributing/deprecation_policy.md
Comment (ajsecord) [inline contributing/deprecation_policy.md:39]: Non-Google contributors need to be able to follow this doc. What would you recommend for them?

File: contributing/deprecation_policy.md
Comment (ajsecord) [inline contributing/deprecation_policy.md]: Pointed out in yesterday's meeting: one still needs to deprecate for non-Google clients.

File: contributing/deprecation_policy.md
Comment (ajsecord) [inline contributing/deprecation_policy.md:79]: You didn't want a macro/new annotation?

**Review by ajsecord — DISMISSED** [review]

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md:79]: Waiting on a new release of https://github.com/jpsim/SourceKitten before we can support the annotations properly in our toolchain.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md:39]: Added a section for non-googler deprecation proposals.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Fixed.

File: contributing/deprecation_policy.md
Comment (ianegordon) [inline contributing/deprecation_policy.md]: I'd prefer an issue to a project.

File: contributing/deprecation_policy.md
Comment (ianegordon) [inline contributing/deprecation_policy.md]: We've 'added' nine steps to our process, which raises concern.

Can we simplify to:
1. If useful, write a migration guide 
1. Annotate API as __deprecated
1. After at least one release and when there are no internal client blockers, remove API & migration guide.
This will cause the next release to be a major version bump.

I'm not sure we need to tell people that we are going to deprecate an API in comments before we actually annotate the API.  I do not expect clients to notice additional comments, but I do expect them to notice the deprecation compiler warnings.

**Review by ianegordon — DISMISSED** [review]

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: I think an issue's probably sufficient as well. Will adjust.

Comment (jverkoey) [reply contributing/deprecation_policy.md]: Filed https://github.com/material-components/material-components-ios/pull/4803 to create a template for this.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: How will clients that miss our announcements know that an API is [edit: to be] deprecated before we mark it as deprecated?

**Comment by jverkoey** [comment]
Hiya @ajsecord, @randallli, @jmdetloff, @galiak11, @andrewoverton, @wenyuzhang666 - I'll be aiming to close this PR by end of day tomorrow. Please find 15-20 minutes to read through the PR and leave feedback before then.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: One point of data to prove the validity of this current process: client teams, when they file bugs related to migrations, will frequently link to the migration guide doc as a means of sharing knowledge about how to perform the migration. This would be harder for clients to do if we solely relied on emails (and emails are difficult to modify in case we need to make changes).

File: contributing/deprecation_policy.md
Comment (galiak11) [inline contributing/deprecation_policy.md:69]: Does it makes sense to have a migration guide template here - similar to #2?

File: contributing/deprecation_policy.md
Comment (galiak11) [inline contributing/deprecation_policy.md:163]: Do you mean to wait after the API is marked "as ready for deletion" - not "as deprecated"?

File: contributing/deprecation_policy.md
Comment (galiak11) [inline contributing/deprecation_policy.md]: Should we try to group "breaking changes" of different APIs together by designating preferred deletion dates and possibly preferred versions for breaking changes? It might make it easier for our client base to plan ahead for these dates/versions.

**Review by galiak11 — APPROVED** [review]

File: contributing/deprecation_policy.md
Comment (galiak11) [inline contributing/deprecation_policy.md:135]: Should we try to group "breaking changes" of different APIs together by designating preferred deletion dates and possibly preferred versions for breaking changes? It might make it easier for our client base to plan ahead for these dates/versions.  We might also want avoid mixing breaking changes with major new functionality in a single release, so it's easier for client apps to rollback breaking changes.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md:69]: Ah! Good idea. I'll draft up a skeleton structure here.

Comment (jverkoey) [reply contributing/deprecation_policy.md:69]: Done. Please take a look again.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md:163]: We only annotate APIs in two manners in this proposal:

- First, as "to be deprecated".
- Second, as "deprecated".

We won't annotate APIs as ready for deletion because we'll simply delete them once they're ready to be deleted.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md:135]: Good topics to consider.

If we follow this process fully, breaking changes will be mitigated for client teams because by the time we ship our major release with the APIs deleted, all of our internal clients will have already migrated off of the APIs and impact will be minimal.

That being said, this is less the case for our external clients. I do think it's fair to consider lumping multiple deletions together into a single release. How do you feel we might coordinate API deprecations in this manner?

**Comment by jverkoey** [comment]
If you haven't reviewed this PR yet I've put a slot on your calendar today so that you may do so:

@ianegordon ianegordon
@ajsecord ajsecord
 @randallli randallli
 @jmdetloff jmdetloff
 @andrewoverton andrewoverton
 @wenyuzhang666 wenyuzhang666

**Review by randallli — APPROVED** [review]

**Review by andrewoverton — APPROVED** [review]

File: contributing/deprecation_policy.md
Comment (fluxminds) [inline contributing/deprecation_policy.md]: Just curiously, what kind of role should we play when our clients (internal/external) are deprecating the API. Should we actively migrate the code for them or should we do any monitoring on the deprecation speed by measuring the internal API usage periodically?

Comment (jverkoey) [reply contributing/deprecation_policy.md]: From my own experience, I find it valuable to perform at least a couple client migrations myself to make sure I have an accurate understanding of the work required. But once bugs have been filed to each of the client teams it should be possible to be hands-off as client teams prioritize and iterate on the work independently (and this means you can get back to working on other things!)

Comment (fluxminds) [reply contributing/deprecation_policy.md]: I see. Thanks for the suggestion!

**Review by fluxminds — APPROVED** [review]

File: contributing/deprecation_policy.md
Comment (ianegordon) [inline contributing/deprecation_policy.md]: super-nit: Consider:
[ ] Add a comment to the API indicating // ToBeDeprecated

To avoid confusion with __annotate(deprecated).

File: contributing/deprecation_policy.md
Comment (ianegordon) [inline contributing/deprecation_policy.md]: Consider:
After the deprecation date, annotate the API as deprecated

**Review by ianegordon — APPROVED** [review]
My concerns about additional process were discussed offline.
I'd prefer __deprecated_msg() over @warning, but I understand the SourceKitten dependency. I wouldn't block on this issue.
Consider comments before Merge.
Thanks you @jverkoey !

**Comment by ianegordon** [comment]
I think we can go ahead an merge this.
All of our active engineers have given it a +1.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Done.

File: contributing/deprecation_policy.md
Comment (jverkoey) [reply contributing/deprecation_policy.md]: Done.

**Comment by jverkoey** [comment]
Thank you for all the feedback gang! Merging this in now.

File: contributing/deprecation_policy.md
Comment (galiak11) [reply contributing/deprecation_policy.md:135]: Perhaps designate certain dates/months during the year as deletion date releases? and include all deprecated features that are ready for that date (deprecation dates can actually go in anytime so they work better with dev schedule).

File: contributing/deprecation_policy.md
Comment (galiak11) [reply contributing/deprecation_policy.md:69]: Not sure where to look... ?

**Pull request merged**
